### DISCUSSION TITLE
Vector cross-compiled for ppc

### DISCUSSION THREAD

**Discussion by bmallya** [vectordotdev/vector#9560] [Q&A] [answered]
Can someone guide me on cross-compiling vector for ppc? Here is what I have
--------------------------
      sh-4.4# uname -a
      Linux a8041bffb0d9 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
      sh-4.4# rustup show
      Default host: x86_64-unknown-linux-gnu
      rustup home:  /root/.rustup
      
      installed toolchains
      --------------------
      
      stable-powerpc-unknown-linux-gnu
      stable-x86_64-unknown-linux-gnu (default)
      1.53.0-x86_64-unknown-linux-gnu
      
      installed targets for active toolchain
      --------------------------------------
      
      powerpc-unknown-linux-gnu
      x86_64-unknown-linux-gnu
      
      active toolchain
      ----------------
      
      1.53.0-x86_64-unknown-linux-gnu (overridden by '/root/ppc/vector/rust-toolchain')
      rustc 1.53.0 (53cb7b09b 2021-06-17)
      
      sh-4.4#        

---------------------------
And with this when I try to build vector providing a target, I run into an error.
    sh-4.4# cargo build --target=powerpc-unknown-linux-gnu
       Compiling proc-macro2 v1.0.26
       Compiling libc v0.2.98
       Compiling syn v1.0.72
       Compiling serde_derive v1.0.126
    error: linking with `cc` failed: exit status: 1
      |
      = note: "cc" "-m64" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-Wl,--as-needed" "-L" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.0.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.1.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.10.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.11.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.12.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.13.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.14.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.15.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.2.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.3.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.4.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.5.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.6.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.7.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.8.rcgu.o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.9.rcgu.o" "-o" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa" "/root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.188xjvth382161et.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/root/ppc/vector/target/debug/deps" "-L" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-b6b48477bfa8c673.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-f560ec02638f7ffe.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-9c8eadb7013c9e0b.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-8b0ec8dbdb85d0bf.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-ba5d5ee707c805d2.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-55166126dbdd5e46.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-c327b365eae3b2f3.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-416439b546a0d033.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-2581188d29552e15.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-da7b2635bfcce6ef.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-09200ed1945e7b2b.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-223ac369b29f5000.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-39562fe6600dd936.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-9b411bb7a19f81b3.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-64ea0581d80339f7.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-b2dbda88b377d685.rlib" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-2a8415a96ed1d7dc.rlib" "-Wl,--end-group" "/root/.rustup/toolchains/1.53.0-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c4d9a5b072ee3191.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc"
      = note: /root/ppc/vector/target/debug/build/serde_derive-2a6d36fc55b2b4fa/build_script_build-2a6d36fc55b2b4fa.build_script_build.9gdh8nj4-cgu.0.rcgu.o: file not recognized: File format not recognized
              collect2: error: ld returned 1 exit status
    
    
    error: aborting due to previous error

**Comment by jszwedko** [comment, marked as answer]
Hi @bmallya !

We use [`cross`](https://github.com/rust-embedded/cross) to cross-compile Vector. We maintain docker images for the cross compilation as they need a few additional dependencies. You can see our current ones here: https://github.com/vectordotdev/vector/tree/master/scripts/cross . You should be able create a similar one for ppc named (`powerpc-unknown-linux-gnu.dockerfile`) and then build it using `make cross-powerpc-unknown-linux-gnu`

https://github.com/vectordotdev/vector/blob/c878c50fd905048d28742443142aad10fad0620f/Makefile#L213-L245

Hopefully this helps! Let us know if you have further questions.

**Reply by bmallya** [reply]
Thanks for the reply @jszwedko. I had compiled vector on a Ubuntu 18.04 container earlier and now am trying to cross compile it for a ppc (and will eventually have to do it for arm as well). 
After installing podman on the above container, I tried to create a binary for arm using cross, when I ran into the following message. This makes me believe that the host system is probably missing something(s) that Vector might need. Is there a min requirement for the host system,  where we can use this cross and build vector for target platform?

      sh-4.4# make cross-image-armv7-unknown-linux-gnueabihf
      podman build \
              --tag vector-cross-env:armv7-unknown-linux-gnueabihf \
              --file scripts/cross/armv7-unknown-linux-gnueabihf.dockerfile \
              scripts/cross
      Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
      Makefile:223: recipe for target 'cross-image-armv7-unknown-linux-gnueabihf' failed
      make: *** [cross-image-armv7-unknown-linux-gnueabihf] Error 125
      sh-4.4#

**Reply by jszwedko** [reply]
Hi @bmallya !

It looks like that is podman complaining about not being able to build n a container. Looking at https://github.com/containers/buildah/issues/867#issuecomment-405670034 it looks like you should either mount `/var/lib/containers` into the build container or use a different storage driver for the build container, like `vfs`.

**Reply by bmallya** [reply]
Thanks again @jszwedko . Let me start looking into that.  Does this way of compiling (make cross) also accept the features of sources/sinks/transforms that I need instead of the whole lot?

**Reply by bmallya** [reply]
I gave this a try on a Ubuntu 20.04 VM. Do I need to also provide the "build" target to make along with "cross-image-x86_64-unknown-linux-gnu" to get the vector binary compiled?

        make cross-image-x86_64-unknown-linux-gnu
        podman build \
	        --tag vector-cross-env:x86_64-unknown-linux-gnu \
	        --file scripts/cross/x86_64-unknown-linux-gnu.dockerfile \
	        scripts/cross
        STEP 1/3: FROM docker.io/rustembedded/cross:x86_64-unknown-linux-gnu
        Trying to pull docker.io/rustembedded/cross:x86_64-unknown-linux-gnu...
        Getting image source signatures
        Copying blob c128e7c4483b done  
        Copying blob 10e22a0ba65d done  
        Copying blob 1d8e9b3cef3d done  
        Copying blob 6d7263e78d09 done  
        Copying blob f91401dbe7ec done  
        Copying blob 2d473b07cdd5 done  
        Copying blob 6144908f2572 done  
        Copying blob e506de543a44 done  
        Copying blob f59909d4c102 done  
        Copying blob 403e130e5d01 done  
        Copying blob 1d91af9520b3 done  
        Copying blob bdaaa09f0c75 done  
        Copying blob c7f1561581ee done  
        Copying config 9a6b90cd98 done  
        Writing manifest to image destination
        Storing signatures
        STEP 2/3: COPY bootstrap-rhel.sh .
        --> 5928a014afe
        STEP 3/3: RUN ./bootstrap-rhel.sh
        Loaded plugins: fastestmirror, ovl
        Determining fastest mirrors
         * base: centos.excellmedia.net
         * extras: centos.excellmedia.net
         * updates: centos.excellmedia.net
        Metadata Cache Created
        Loaded plugins: fastestmirror, ovl
        Loading mirror speeds from cached hostfile
         * base: centos.excellmedia.net
         * extras: centos.excellmedia.net
         * updates: centos.excellmedia.net
        Resolving Dependencies
        --> Running transaction check
        ---> Package clang.x86_64 0:3.4.2-8.el7 will be installed
        --> Processing Dependency: libLLVM-3.4.so()(64bit) for package: clang-3.4.2-8.el7.x86_64
        ---> Package llvm.x86_64 0:3.4.2-8.el7 will be installed
        --> Running transaction check
        ---> Package llvm-libs.x86_64 0:3.4.2-8.el7 will be installed
        --> Finished Dependency Resolution
        
        Dependencies Resolved
        
        ================================================================================
         Package            Arch            Version               Repository       Size
        ================================================================================
        Installing:
         clang              x86_64          3.4.2-8.el7           extras           19 M
         llvm               x86_64          3.4.2-8.el7           extras          1.3 M
        Installing for dependencies:
         llvm-libs          x86_64          3.4.2-8.el7           extras          7.6 M
        
        Transaction Summary
        ================================================================================
        Install  2 Packages (+1 Dependent package)
        
        Total download size: 28 M
        Installed size: 93 M
        Downloading packages:
        --------------------------------------------------------------------------------
        Total                                              5.9 MB/s |  28 MB  00:04     
        Running transaction check
        Running transaction test
        Transaction test succeeded
        Running transaction
          Installing : llvm-libs-3.4.2-8.el7.x86_64                                 1/3 
          Installing : llvm-3.4.2-8.el7.x86_64                                      2/3 
          Installing : clang-3.4.2-8.el7.x86_64                                     3/3 
          Verifying  : clang-3.4.2-8.el7.x86_64                                     1/3 
          Verifying  : llvm-libs-3.4.2-8.el7.x86_64                                 2/3 
          Verifying  : llvm-3.4.2-8.el7.x86_64                                      3/3 
        
        Installed:
          clang.x86_64 0:3.4.2-8.el7              llvm.x86_64 0:3.4.2-8.el7             
        
        Dependency Installed:
          llvm-libs.x86_64 0:3.4.2-8.el7                                                
        
        Complete!
        COMMIT vector-cross-env:x86_64-unknown-linux-gnu
        --> eedd00ba84e
        Successfully tagged localhost/vector-cross-env:x86_64-unknown-linux-gnu
        eedd00ba84eb11a872f9097591e6e8cbb6fb439d214125489b0d718db189e333

When I tried the option that you mentioned, I see an  error popping up [I am using v0.15.1 of Vector]
      make cross-x86_64-unknown-linux-gnu
      make -k cross-image-unknown-linux-gnu
      make[1]: Entering directory '/home/user/bm/vector'
      podman build \
	      --tag vector-cross-env:unknown-linux-gnu \
	      --file scripts/cross/unknown-linux-gnu.dockerfile \
	      scripts/cross
      Error: stat /home/user/bm/vector/scripts/cross/scripts/cross/unknown-linux-gnu.dockerfile: no such file or directory
      make[1]: *** [Makefile:223: cross-image-unknown-linux-gnu] Error 125
      make[1]: Leaving directory '/home/user/bm/vector'
      make: *** [Makefile:237: cross-x86_64-unknown-linux-gnu] Error 2

**Reply by jszwedko** [reply]
Hi @bmallya !

You are trying to cross compile for ppc right? I ask because the above commands are referencing other platforms like `x86_64-unknown-linux-gnu`.

If yes, I think you'll need to:

1. Add a new dockerfile at `scripts/cross/powerpc-unknown-linux-gnu.dockerfile`. It should be something like:
```
FROM rustembedded/cross:powerpc-unknown-linux-gnu

COPY bootstrap-ubuntu.sh .
RUN ./bootstrap-ubuntu.sh
```
2. Run `make cross-build-powerpc-unknown-linux-gnu`

That should build the binary and put it in `target/powerpc-unknown-linux-gnu/`

**Reply by jszwedko** [reply]
Ah, yes, you are right, I forgot to include `-build` in the above command I provided the first time.

**Reply by jszwedko** [reply]
And I forgot it needs to be added to `Cargo.toml` too 😄 I opened https://github.com/vectordotdev/vector/compare/add-powerpc-build to add these files to the repo.

However, it looks like the `metrics-rs` library does not currently support PowerPC due to lack of support for 64-bit atomic integers meaning I don't think Vector can be build on PowerPC currently 😢 We would need upstream support first.

```
 --> /cargo/registry/src/github.com-1ecc6299db9ec823/metrics-0.17.0/src/key.rs:7:32
  |
7 |     sync::atomic::{AtomicBool, AtomicU64, Ordering},
  |                                ^^^^^^^^^
  |                                |
  |                                no `AtomicU64` in `sync::atomic`
  |                                help: a similar name exists in the module: `AtomicU8`
```

**Reply by jszwedko** [reply]
I opened https://github.com/metrics-rs/metrics/issues/235 upstream.

**Reply by bmallya** [reply]
Thanks for the help @jszwedko.
To answer your question, yes, I want Vector to be compiled for ppc. But since I had not used the cross option yet, I wanted to give a try using the existing dockerfile(s).
Could you also tell me if I its supported to pass only certain components that I need, as you had helped me earlier in this question [Vector Lite](https://github.com/vectordotdev/vector/discussions/9010)

**Reply by bmallya** [reply]
@jszwedko , is there also any change in the .cargo/config file to be done for powerpc? When I added a new docker file for powerpc along with the change in the Cargo.toml file, I ran into an error

          cross build \
	          --release \
	          --target powerpc-unknown-linux-gnu \
	          --no-default-features \
	          --features target-powerpc-unknown-linux-gnu
          info: syncing channel updates for '1.53.0-x86_64-unknown-linux-gnu'
          
            1.53.0-x86_64-unknown-linux-gnu unchanged - rustc 1.53.0 (53cb7b09b 2021-06-17)
          
          info: checking for self-updates
             Compiling ring v0.16.20
          error: failed to run custom build command for `ring v0.16.20`
          
          Caused by:
            process didn't exit successfully: `/target/release/build/ring-38cfb14ce7ad78d3/build-script-build` (exit status: 101)
            --- stderr
            thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/build.rs:358:10
            note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
          make: *** [Makefile:238: cross-build-powerpc-unknown-linux-gnu] Error 101

**Reply by jszwedko** [reply]
Hi @bmallya !

Which version of Vector are you trying to build? Could you rerun that with `RUST_BACKTRACE=1` and provide the full stack-trace?

> Could you also tell me if I its supported to pass only certain components that I need, as you had helped me earlier in this question Vector Lite

You can if you run `cross` directly as it seems like you are. Something like:

```
  cross build \
          --release \
          --target powerpc-unknown-linux-gnu \
          --no-default-features \
          --features sources-stdin,sinks-console,transforms-remap
```

This would create a build with just the specified components.

Unfortunately, `metrics-rs` is a core dependency of Vector that I don't think you'll be able to build without so I believe you'll eventually run into the same issue I hit above where `metrics-rs` doesn't support non-64-bit systems. We'll need to patch it and other dependencies before it'll be possible to build Vector for 32-bit PowerPC. If you are a Rust developer and wanted to help with that effort it does look like `metrics-rs` is open to making the necessary changes to be backwards compatible: https://github.com/metrics-rs/metrics/issues/235#issuecomment-944257430

**Reply by jszwedko** [reply]
Actually, I see, `ring`, a transitive dependency, doesn't support PowerPC as a target yet either: https://github.com/briansmith/ring/issues/389

**Reply by bmallya** [reply]
Hello @jszwedko, I have been using v0.15.1 so far. Would you suggest to move to a newer version?

I am not a RUST developer yet, but I will start looking at what has been proposed for metrics-rs. And I noticed that there is some work going on for "ring", but not sure when it would be done

**Reply by jszwedko** [reply]
I was just curious which version so I could try building that one specifically. I would recommend updating just to keep current, but no other particular reason.

👍 if you wanted to help push dependencies forward with support for PPC that would definitely move us closer to being able to support it is a build target.

**Reply by bmallya** [reply]
@jszwedko, Would you know which library/crate is depending on ring? I was wondering if I can build without that dependency. 

Another thing that I probably wanted to run by you and get your suggestion. I made a small change in the Makefile to accept the target features  and ran the following command

        make cross-build-powerpc-unknown-linux-gnu LRU="sources-exec,sources-file,sources-generator,sources-http,sources-internal_logs,sources-socket,sources-stdin,sources-syslog,sources-vector,sources-internal_metrics,sources-host_metrics,transforms-add_fields,transforms-ansi_stripper, transforms-coercer,transforms-concat,transforms-dedupe,transforms-field_filter,transforms-filter,transforms-json_parser,transforms-key_value_parser,transforms-log_to_metric,transforms-logfmt_parser,transforms-merge,transforms-metric_to_log,transforms-reduce,transforms-regex_parser,transforms-remap,transforms-remove_fields,transforms-rename_fields,transforms-route,transforms-sample,transforms-split,transforms-tokenizer,sinks-blackhole,sinks-console,sinks-file,sinks-http,sinks-socket,sinks-vector,vendor-all"

which threw up,
      
      Processing triggers for systemd (229-4ubuntu21.31) ...
      COMMIT vector-cross-env:powerpc-unknown-linux-gnu
      --> 63a2fda71ae
      Successfully tagged localhost/vector-cross-env:powerpc-unknown-linux-gnu
      63a2fda71ae73df31517a1ea3849d908bc6c6054cc61cebde00302c57b795622
      make[1]: Leaving directory '/home/bmallya/Platforms/vector'
      cross build \
	      --release \
	      --target powerpc-unknown-linux-gnu \
	      --no-default-features \
      #		--features target-powerpc-unknown-linux-gnu
      info: syncing channel updates for '1.55.0-x86_64-unknown-linux-gnu'
      
        1.55.0-x86_64-unknown-linux-gnu unchanged - rustc 1.55.0 (c8dfcfe04 2021-09-06)
      
      info: checking for self-updates
         Compiling openssl-sys v0.9.66
      error: failed to run custom build command for `openssl-sys v0.9.66`
      
      Caused by:
        process didn't exit successfully: `/target/release/build/openssl-sys-fd342e47d78f02e6/build-script-main` (exit status: 101)
        --- stdout
        cargo:rustc-cfg=const_fn
        cargo:rerun-if-env-changed=POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
        POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR unset
        cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
        OPENSSL_LIB_DIR unset
        cargo:rerun-if-env-changed=POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
        POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR unset
        cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
        OPENSSL_INCLUDE_DIR unset
        cargo:rerun-if-env-changed=POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_DIR
        POWERPC_UNKNOWN_LINUX_GNU_OPENSSL_DIR unset
        cargo:rerun-if-env-changed=OPENSSL_DIR
        OPENSSL_DIR unset
        cargo:rerun-if-env-changed=OPENSSL_NO_PKG_CONFIG
        cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_powerpc-unknown-linux-gnu
        cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS_powerpc_unknown_linux_gnu
        cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_ALLOW_CROSS
        cargo:rerun-if-env-changed=PKG_CONFIG_ALLOW_CROSS
        cargo:rerun-if-env-changed=PKG_CONFIG
        cargo:rerun-if-env-changed=OPENSSL_STATIC
        cargo:rerun-if-env-changed=OPENSSL_DYNAMIC
        cargo:rerun-if-env-changed=PKG_CONFIG_ALL_STATIC
        cargo:rerun-if-env-changed=PKG_CONFIG_ALL_DYNAMIC
        cargo:rerun-if-env-changed=PKG_CONFIG_PATH_powerpc-unknown-linux-gnu
        cargo:rerun-if-env-changed=PKG_CONFIG_PATH_powerpc_unknown_linux_gnu
        cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_PATH
        cargo:rerun-if-env-changed=PKG_CONFIG_PATH
        cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_powerpc-unknown-linux-gnu
        cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_powerpc_unknown_linux_gnu
        cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_LIBDIR
        cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
        cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_powerpc-unknown-linux-gnu
        cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_powerpc_unknown_linux_gnu
        cargo:rerun-if-env-changed=TARGET_PKG_CONFIG_SYSROOT_DIR
        cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
        run pkg_config fail: "`\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"` did not exit successfully: exit status: 1\n--- stderr\nPackage openssl was not found in the pkg-config search path.\nPerhaps you should add the directory containing `openssl.pc'\nto the PKG_CONFIG_PATH environment variable\nNo package 'openssl' found\n"

I tried installing the libssl package into the container through the boot-ubuntu.sh script, but that did not help either

**Reply by jszwedko** [reply]
Nice! I imagine we'd be happy to accept the Makefile enhancement if you want to open a PR.

I think you might need to install `libssl-dev`? That would have Vector dynamically link with it. You can also have it statically link against a vendored openssl via adding `vendor-openssl` to your list of features.

Unfortunately there are quite a number of dependencies on `ring`:

```
$ cargo tree -i -p ring
ring v0.16.20
├── azure_storage v0.1.0 (https://github.com/Azure/azure-sdk-for-rust.git?rev=16bcf0ab1bb6e380d966a69d314de1e99ede553a#16bcf0ab)
│   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   [dev-dependencies]
│   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
├── jsonwebtoken v7.2.0
│   └── gouth v0.2.1
│       └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
├── rustls v0.18.1
│   └── attohttpc v0.17.0
│       └── gouth v0.2.1 (*)
├── rustls v0.19.1
│   ├── bollard v0.11.0
│   │   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   ├── hyper-rustls v0.22.1
│   │   ├── bollard v0.11.0 (*)
│   │   └── reqwest v0.11.6
│   │       ├── azure_core v0.1.0 (https://github.com/Azure/azure-sdk-for-rust.git?rev=16bcf0ab1bb6e380d966a69d314de1e99ede553a#16bcf0ab)
│   │       │   ├── azure_storage v0.1.0 (https://github.com/Azure/azure-sdk-for-rust.git?rev=16bcf0ab1bb6e380d966a69d314de1e99ede553a#16bcf0ab) (*)
│   │       │   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   │       │   [dev-dependencies]
│   │       │   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   │       ├── goauth v0.10.0
│   │       │   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   │       ├── oauth2 v4.1.0
│   │       │   └── azure_core v0.1.0 (https://github.com/Azure/azure-sdk-for-rust.git?rev=16bcf0ab1bb6e380d966a69d314de1e99ede553a#16bcf0ab) (*)
│   │       ├── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   │       └── vector-api-client v0.1.2 (/Users/jesse.szwedko/workspace/vector/lib/vector-api-client)
│   │           └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   │       [dev-dependencies]
│   │       └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   ├── mongodb v2.0.1
│   │   └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   ├── nats v0.10.1
│   │   └── async-nats v0.10.1
│   │       └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
│   ├── reqwest v0.11.6 (*)
│   ├── rustls-native-certs v0.5.0
│   │   ├── bollard v0.11.0 (*)
│   │   ├── hyper-rustls v0.22.1 (*)
│   │   └── nats v0.10.1 (*)
│   └── tokio-rustls v0.22.0
│       ├── hyper-rustls v0.22.1 (*)
│       ├── mongodb v2.0.1 (*)
│       ├── reqwest v0.11.6 (*)
│       └── tonic v0.5.2
│           └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
├── sct v0.6.1
│   ├── ct-logs v0.8.0
│   │   ├── bollard v0.11.0 (*)
│   │   └── hyper-rustls v0.22.1 (*)
│   ├── rustls v0.18.1 (*)
│   └── rustls v0.19.1 (*)
└── webpki v0.21.4
    ├── attohttpc v0.17.0 (*)
    ├── hyper-rustls v0.22.1 (*)
    ├── mongodb v2.0.1 (*)
    ├── nats v0.10.1 (*)
    ├── rustls v0.18.1 (*)
    ├── rustls v0.19.1 (*)
    ├── tokio-rustls v0.22.0 (*)
    ├── webpki-roots v0.19.0
    │   └── attohttpc v0.17.0 (*)
    └── webpki-roots v0.21.1
        ├── bollard v0.11.0 (*)
        ├── mongodb v2.0.1 (*)
        └── reqwest v0.11.6 (*)
```

**Reply by jszwedko** [reply]
Actually! There are only a couple of deps when using your feature set:

```
$ cargo tree -i -p ring --no-default-features --features 'sources-exec,sources-file,sources-generator,sources-http,sources-internal_logs,sources-socket,sources-stdin,sources-syslog,sources-vector,sources-internal_metrics,sources-host_metrics,transforms-add_fields,transforms-ansi_stripper, transforms-coercer,transforms-concat,transforms-dedupe,transforms-field_filter,transforms-filter,transforms-json_parser,transforms-key_value_parser,transforms-log_to_metric,transforms-logfmt_parser,transforms-merge,transforms-metric_to_log,transforms-reduce,transforms-regex_parser,transforms-remap,transforms-remove_fields,transforms-rename_fields,transforms-route,transforms-sample,transforms-split,transforms-tokenizer,sinks-blackhole,sinks-console,sinks-file,sinks-http,sinks-socket,sinks-vector,vendor-all' --edges no-dev
ring v0.16.20
├── rustls v0.19.1
│   └── tokio-rustls v0.22.0
│       └── tonic v0.5.2
│           └── vector v0.17.2 (/Users/jesse.szwedko/workspace/vector)
├── sct v0.6.1
│   └── rustls v0.19.1 (*)
└── webpki v0.21.4
    ├── rustls v0.19.1 (*)
    └── tokio-rustls v0.22.0 (*)
```

Basically it looks like it's just the `vector` source/sink since they depend `tonic`. If I remove those two components, there are no dependencies on `ring`.

**Reply by bmallya** [reply]
Thanks for the quick response @jszwedko. Really appreciate all the help you are providing me.
On the open-ssl error that I mentioned earlier, I had libssl-dev added into the boot-ubuntu.sh script
         
      apt-get update
        apt-get install -y \
          apt-transport-https \
          libssl-dev \
          wget
        
        # we need LLVM >= 3.9 for onig_sys/bindgen
        
        cat <<-EOF > /etc/apt/sources.list.d/llvm.list
        deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-9 main
        deb-src http://apt.llvm.org/xenial/ llvm-toolchain-xenial-9 main
        EOF
        
        wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key| apt-key add -
        
        apt-get update
        
        # needed by onig_sys
        apt-get install -y \
              libclang1-9 \
              llvm-9

 In the features that I had passed to the makefile, **vendor-all**  is also present which should also mean **vendor-openssl**  is also in. Correct?

Unfortunately, **vector** as a source/sink is at the heart of the solution we are using Vector for. The intention is to have like a vendor agent running on some systems in the network. These agents would collect the logs from different sources (say http/tcp/socket) and then these are forwarded to a centralised collector (on a different system). And the way the logs are forwarded is to use **vector** as a sink (on the agent) and source( on the collector)

**Reply by jszwedko** [reply]
Hey! Apologies for the delayed response. Correct `vendor-all` will include `vendor-openssl`. I'm not sure why you'd be seeing those OpenSSL issues.

You could try using another source/sink pair to send data. Other users have used the `splunk_hec` source and sink to send logs between Vector instances. It looks like that would remove any dependency on `ring`.
